### PR TITLE
Add commonly used data structures to builtin

### PR DESCRIPTION
**Pull request by HybridDog**
Implementation of #7490

### CODE REVIEW

**Comment by raymoo** [comment]
Summary of worst case time complexity of possible implementations:

Old code (n = number of jobs in queue, k = number of ready jobs):
- Add a job: O(1)
- Run and remove all ready jobs: O(kn), which is potentially O(n^2)

This code:
- Add a job: O(log n)
- Run and remove all ready jobs: O(k log n), which is potentially O(n log n)

Linear-scanning priority queue that fixes the performance bug in the current implementation:
- Add a job: O(1)
- Run and remove all ready jobs: O(n)

So depending on the actual workload, the binary heap may have better or worse performance than a linear-scanning priority queue implementation that is better than the current one, but the difference in practice will be small because there are normally not that many jobs in the queue (less than a thousand).

**Comment by HybridDog** [comment]
I assume that the globalstep is executed a lot more frequently than adding or removing a job.
With linear-scanning priority queue searching for an expired job is O(n), whereas with the binary heap it is O(1).
If one job gets added and executed, it's O(1 + n) = O(n) with linear-scanning and O(log(n) + log(n)) = O(log(n)) with the binary heap.
Anyway, the commit is just an example usage of a provided priority queue data structure.

File: builtin/common/datastructures.lua [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+				return v
+			end
+			-- source is empty, swap it with sink
+			self.source, self.sink = self.sink, self.source
+			self.n_out = self.n_in
+			self.n_in = 0
+			local v = self.source[1]
+			self.source[1] = nil
+			self.p_out = 2
+			return v
```
Comment (DS-Minetest) [inline builtin/common/datastructures.lua]: You could swap the uppermost element with the current one and set the uppermost one to `nil` and when you reach the top you can make the source empty from above.

File: builtin/common/datastructures.lua [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+			for i = 1, self.n do
+				t[i] = value_tostring(self[i])
+			end
+			return self.n .. " elements; bottom to top: " ..
+				table.concat(t, ", ")
+		end,
+	}
+}
+
+function core.create_stack(data)
```
Comment (DS-Minetest) [inline builtin/common/datastructures.lua]: Why not `input, n` and then set `n = n or input.n`? Wouldn't that be more efficient?

Comment (HybridDog) [reply builtin/common/datastructures.lua]: core.create_stack does not need to be as efficient as possible, it's only invoked when creating a stack, an extendable parameter list passed as table is more desirable.

File: builtin/common/datastructures.lua (hunk 1 above)
Comment (HybridDog) [reply builtin/common/datastructures.lua]: This doesn't improve performance, does it?

Comment (DS-Minetest) [reply builtin/common/datastructures.lua]: It could maybe save memory.
It was probably a bad suggestion. >_<

**Comment by HybridDog** [comment]
I've removed the usage examples because I put them only for demonstration purposes.

**Comment by SmallJoker** [comment]
Even though these are pretty basic operations, wouldn't a library mod fit better? I think only a few mods could actually make use of these structures.

**Comment by HybridDog** [comment]
It is possible to copy-paste the datastructures code to a mod (or use git subtree), so I don't know if they are important enough to be implemented directly in minetest. You can close this if a mod (or copied code) fits better.

**Comment by SmallJoker** [comment]
The code is there. Contributors may pick it up at any time they like.
For now this PR first best into a (library) mod.

**Pull request closed without merging**